### DISCUSSION TITLE
createSignedUrl - 404

### DISCUSSION THREAD

**Discussion by Hallidayo** [supabase/supabase#5601] [Questions] [answered]
Hi,

I'm trying to create a signed URL for an image that I have stored in one of my buckets in a nextjs application with supabase auth. The bucket itself is a private bucket and I have the following policy created on the bucket for `SELECT`, `UPDATE` AND `UPDATE` : 

`((bucket_id = 'images_bucket'::text) AND ((uid())::text = (storage.foldername(name))[1]))`

When I use the createSignedUrl function in getServerSideProps the following error object is returned:

`{ message: 'The resource was not found', status: 400 }`

I've checked I am logged in with a authenticated user and when I use `supabase.auth.api.getUserByCookie(req, res)` I get a valid response back with the user details

When I switch off my polices and make the bucket public and use `getPublicUrl` instead than I can get a link and the link displays the image.

What could be causing the 400 error when using createSignedUrl?

Thankyou.

**Comment by GaryAustin1** [comment]
If your bucket structure is images_bucket/uid/filename then that policy looks correct.  If you have not yet try hardcoding the uid in the policy.  If that works then you are not actually sending a token with a user ID in it from the server.

getPublicUrl is not a good test of anything as it does not make a database call at all and just builds up the url from constants (well fixed values for your instance) and your pathname.

**Reply by Hallidayo** [reply]
Thankyou for your reply @GaryAustin1.

As you have suggested, I've hardcoded the uid in my policy and this is generating a URL with a token for the image and the image is viewable. So this would mean the policy in place is correct?

How would I go about sending the user ID in a token to make a request as when I check the cookies on the page I have a active cookie set named `sb:token` which seems to have the user information?

**Reply by GaryAustin1** [reply]
So your policy is good.
I'm guessing you are running this code while on the server?  I'm really not up to speed on all the issues with frame works on servers and auth, but people seem to have issues getting it sorted out, especially getting the proper token for api requests.

You could put a basic table select where you make the storage call and try true/authorized/uid select policy and see what works and doesn't.

Edit: this is close to a hint as I can give you on server set up and may not even be relevant to your issue.:
https://supabase.com/docs/reference/javascript/auth-setauth

You might be better off framing a new question with using next.js (describe your environment where you are making the call) and say it does not look proper token for user is getting to supabase.  This would be easier if you have the simple select on table with policy just so time is not wasted on figuring storage policy differences.

**Reply by Hallidayo** [reply]
Yeah Im using the `getServerSideProps` function that is provided by next.js to generate the images which I am doing before the page renders.

Just for reference my stack is:

Supabase
Next JS
Prisma

When I am trying to do a `SELECT` statement for testing via supabase I am actually getting the following error object:

`  error: {
    hint: null,
    details: null,
    code: '42501',
    message: 'permission denied for schema public'
  }`

Having had a look around though I believe this issue is related to Prisma being used with the supabase database.

**Comment by Hallidayo** [comment, marked as answer]
@GaryAustin1 - Again Gary thankyou for your help on the above.

I managed to solve my issue on my project by instead of generating the signed URL on the server side I changed to generate the signedURL on the client side. I have a function which does the following:

```
  const getImage = async (image) => {
    try {
      const { signedURL, error } = await supabase.storage
        .from(<folder name>)
        .createSignedUrl(<file>, 240);
      if (error) {
        throw error;
      }
      setImageUrl(signedURL);
    } catch (error) {
      console.log("Error: ", error.message);
    }
  };
```
Then the above function then gets call in a `useEffect` hook with image being a prop:
```
  useEffect(() => {
    if (image) {
      getImage(image);
    }
  }, []);
```
So after a successful download of the image the URL is then saved as local state which will then render the image. This seems to be working nicely with no performance issues so far.
